File: utils/logger.py
"""Logging utility module."""
import logging
import os
from datetime import datetime


class Log:
    """Logger class for the automation framework."""

    _logger = None

    @classmethod
    def _get_logger(cls):
        """Get or create logger instance."""
        if cls._logger is None:
            cls._logger = logging.getLogger("AutomationFramework")
            cls._logger.setLevel(logging.INFO)

            # Create logs directory if it doesn't exist
            if not os.path.exists("logs"):
                os.makedirs("logs")

            # Console handler
            console_handler = logging.StreamHandler()
            console_handler.setLevel(logging.INFO)

            # File handler
            log_file = f"logs/test_execution.log"
            file_handler = logging.FileHandler(log_file, mode='a', encoding='utf-8')
            file_handler.setLevel(logging.INFO)

            # Formatter
            formatter = logging.Formatter(
                '%(asctime)s [%(threadName)s] %(levelname)-5s %(name)s - %(message)s',
                datefmt='%Y-%m-%d %H:%M:%S'
            )
            console_handler.setFormatter(formatter)
            file_handler.setFormatter(formatter)

            # Add handlers
            cls._logger.addHandler(console_handler)
            cls._logger.addHandler(file_handler)

        return cls._logger

    @classmethod
    def info(cls, message: str):
        """Log info message."""
        cls._get_logger().info(message)

    @classmethod
    def error(cls, message: str):
        """Log error message."""
        cls._get_logger().error(message)

    @classmethod
    def warn(cls, message: str):
        """Log warning message."""
        cls._get_logger().warning(message)

    @classmethod
    def debug(cls, message: str):
        """Log debug message."""
        cls._get_logger().debug(message)

File: hybrid_typescript_converted/utils/logger.ts
/**
 * Logger Utility
 * Centralized logging with file and console output
 * Migrated from utils/logger.py
 */

import * as winston from 'winston';
import * as path from 'path';
import * as fs from 'fs';

export class Logger {
  private static instance: winston.Logger | null = null;

  /**
   * Get or create logger instance (Singleton pattern)
   */
  private static getLogger(): winston.Logger {
    if (!this.instance) {
      // Ensure logs directory exists
      const logsDir = path.join(process.cwd(), 'logs');
      if (!fs.existsSync(logsDir)) {
        fs.mkdirSync(logsDir, { recursive: true });
      }

      // Define log format
      const logFormat = winston.format.combine(
        winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
        winston.format.printf(({ timestamp, level, message }) => {
          return `${timestamp} [${process.pid}] ${level.toUpperCase().padEnd(5)} AutomationFramework - ${message}`;
        })
      );

      // Create logger instance
      this.instance = winston.createLogger({
        level: process.env.LOG_LEVEL || 'info',
        format: logFormat,
        transports: [
          // Console transport
          new winston.transports.Console({
            format: winston.format.combine(
              winston.format.colorize(),
              logFormat
            ),
          }),
          // File transport
          new winston.transports.File({
            filename: path.join(logsDir, 'test-execution.log'),
            format: logFormat,
            options: { flags: 'a' }, // Append mode
          }),
        ],
      });
    }

    return this.instance;
  }

  /**
   * Log info message
   */
  static info(message: string): void {
    this.getLogger().info(message);
  }

  /**
   * Log error message
   */
  static error(message: string): void {
    this.getLogger().error(message);
  }

  /**
   * Log warning message
   */
  static warn(message: string): void {
    this.getLogger().warn(message);
  }

  /**
   * Log debug message
   */
  static debug(message: string): void {
    this.getLogger().debug(message);
  }

  /**
   * Close logger and flush
   */
  static close(): void {
    if (this.instance) {
      this.instance.close();
      this.instance = null;
    }
  }
}

// Export singleton instance methods
export const Log = Logger;
